### PR TITLE
[core] improve the performance of the get object location by batching

### PR DESCRIPTION
**Pull request by Bye-legumes**
## Why are these changes needed?

In our application with @nemo9cby @liuxsh9, which involves nested remote calls, we've noticed the query object to the driver may result high pressure for the driver process and results in the down of the driver node. Similar issue can be found at  https://github.com/ray-project/ray/issues/43466

In this PR, we change `GetLocationFromOwner` so it will group by the owner address then send batch of ojbect_ids to query instead set up rpc for each object_id. Mostly the owner is the driver so it can reduce the pressure of the driver.

## Related issue number
https://github.com/ray-project/ray/issues/43466

## Checks

- [√] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [√] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [√] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [√] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by Bye-legumes** [comment]
I believe generally this implementation is correct but for the error handling part if you have any suggestions I can also make modification. Their are no influences for the correct status cases, but if there are errors, currently I add warning in the handler at object_id level and keep the warning in the callback. Do you think this is enough or you have better suggestions?

**Comment by raulchen** [comment]
@jjyao can you find someone from the Core team to take a look at this PR? 
This feature is useful for Data. We had to disable locality because of bad performance. Would like to see if this can PR.
@Bye-legumes if you also have any benchmark results with this PR, that will be very helpful.

File: src/ray/protobuf/core_worker.proto [hunk 1]
```diff
@@ -226,13 +226,15 @@ message ObjectLocationUpdate {
 }
 
 message GetObjectLocationsOwnerRequest {
-  WorkerObjectLocationsSubMessage object_location_request = 1;
+  bytes intended_worker_id = 1;
+  repeated bytes object_id = 2;
```
Comment (jjyao) [inline src/ray/protobuf/core_worker.proto:230]: object_ids

File: src/ray/core_worker/core_worker.cc [hunk 2]
```diff
@@ -3773,16 +3804,24 @@ void CoreWorker::HandleGetObjectLocationsOwner(
...
-  auto object_info = reply->mutable_object_location_info();
-  auto status = reference_counter_->FillObjectInformation(object_id, object_info);
-  send_reply_callback(status, nullptr, nullptr);
+  Status overall_status = Status::OK();  // Assume success initially
+  for (int i = 0; i < request.object_id_size(); ++i) {
+    auto object_id = ObjectID::FromBinary(request.object_id(i));
+    auto object_info = reply->add_object_location_infos();
+    // TODO(zhilong): Need to deal with fail query for each object?
+    auto current_status =
+        reference_counter_->FillObjectInformation(object_id, object_info);
```
Comment (jjyao) [inline src/ray/core_worker/core_worker.cc:3817]: I checked `FillObjectInformation()`, it always return Status::OK(). Could you change the method signature to return void so the caller doesn't need to handle error.

File: src/ray/core_worker/core_worker.cc [hunk 3]
```diff
@@ -1783,31 +1794,51 @@ Status CoreWorker::GetLocationFromOwner(
...
   auto ready_promise = std::make_shared<std::promise<void>>();
   auto location_by_id =
       std::make_shared<absl::flat_hash_map<ObjectID, std::shared_ptr<ObjectLocation>>>();
 
-  for (const auto &object_id : object_ids) {
-    rpc::Address owner_address;
-    RAY_RETURN_NOT_OK(GetOwnerAddress(object_id, &owner_address));
+  for (const auto &pair : objects_by_owner) {
+    const auto &owner_address = pair.first;
+    const auto &owner_object_ids = pair.second;
```
Comment (jjyao) [inline src/ray/core_worker/core_worker.cc:1812]: I think you can do

```
for (const auto &[owner_address, owner_object_ids]: objects_by_owner)
```

File: src/ray/core_worker/core_worker.cc [hunk 4]
```diff
@@ -1783,31 +1794,51 @@ Status CoreWorker::GetLocationFromOwner(
...
+              // Assuming successful query, emplace the result
+              location_by_id->emplace(owner_object_ids[i],
+                                      std::make_shared<ObjectLocation>(locations[i]));
+            }
           } else {
-            RAY_LOG(WARNING) << "Failed to query location information for " << object_id
-                             << " with error: " << status.ToString();
+            RAY_LOG(WARNING)
+                << "Failed to query location information for objects owned by "
+                << owner_address.worker_id() << " with error: " << status.ToString();
```
Comment (jjyao) [inline src/ray/core_worker/core_worker.cc:1841]: << "Failed to query location information for objects " << debug_string(owner_object_ids) << " owned by " << ....

File: src/ray/core_worker/core_worker.cc [hunk 5]
```diff
@@ -1783,31 +1794,51 @@ Status CoreWorker::GetLocationFromOwner(
...
+                        const rpc::GetObjectLocationsOwnerReply &reply) {
           absl::MutexLock lock(mutex.get());
           if (status.ok()) {
-            location_by_id->emplace(
-                object_id, std::make_shared<ObjectLocation>(CreateObjectLocation(reply)));
+            auto locations = CreateObjectLocations(reply);
+            for (size_t i = 0; i < owner_object_ids.size(); ++i) {
+              // Assuming successful query, emplace the result
+              location_by_id->emplace(owner_object_ids[i],
+                                      std::make_shared<ObjectLocation>(locations[i]));
```
Comment (jjyao) [inline src/ray/core_worker/core_worker.cc:1836]: We can just do 

```
location_by_id->emplace(owner_object_ids[i],
                                      std::make_shared<ObjectLocation>(CreateObjectLocation(reply.object_location_info(i)));
```

So we don't need the extra `GetObjectLocations()` method.

File: src/ray/core_worker/core_worker.cc [hunk 6]
```diff
@@ -1783,31 +1794,51 @@ Status CoreWorker::GetLocationFromOwner(
...
       std::make_shared<absl::flat_hash_map<ObjectID, std::shared_ptr<ObjectLocation>>>();
 
-  for (const auto &object_id : object_ids) {
-    rpc::Address owner_address;
-    RAY_RETURN_NOT_OK(GetOwnerAddress(object_id, &owner_address));
+  for (const auto &pair : objects_by_owner) {
+    const auto &owner_address = pair.first;
+    const auto &owner_object_ids = pair.second;
     auto client = core_worker_client_pool_->GetOrConnect(owner_address);
     rpc::GetObjectLocationsOwnerRequest request;
```
Comment (jjyao) [inline src/ray/core_worker/core_worker.cc:1814]: Currently Ray grpc max message size is 512 MB. In theory, if we have lots of objects for a single owner, the request size can exceeds that limit. Although it should be very rare, can we set a batch size limit just to be safe: e.g. a single batch can have at most 100 objects.

**Review by jjyao — COMMENTED** [review]
LG!!

File: src/ray/protobuf/core_worker.proto (hunk 1 above)
Comment (Bye-legumes) [reply src/ray/protobuf/core_worker.proto]: fixed!

File: src/ray/core_worker/core_worker.cc (hunk 2 above)
Comment (Bye-legumes) [reply src/ray/core_worker/core_worker.cc]: I just change FillObjectInformation from Status to OK, and also make some modifications to the test

File: src/ray/core_worker/core_worker.cc (hunk 3 above)
Comment (Bye-legumes) [reply src/ray/core_worker/core_worker.cc:1800]: This is my original implementation but failed at CI, it seems that Structured bindings (introduced in C++17) cannot be directly captured by a lambda until C++20, and even then. So I move it back to define them explicitly.
error from ci 
```
[2024-03-26T12:45:33Z] #13 160.1     src/ray/core_worker/core_worker.cc:1821:6: error: 'owner_object_ids' in capture list does not name a variable
[2024-03-26T12:45:33Z] #13 160.1         [owner_object_ids, mutex, num_remaining, ready_promise, location_by_id, this, owner_address](
[2024-03-26T12:45:33Z] #13 160.1          ^
[2024-03-26T12:45:33Z] #13 160.1     src/ray/core_worker/core_worker.cc:1821:83: error: 'owner_address' in capture list does not name a variable
[2024-03-26T12:45:33Z] #13 160.1         [owner_object_ids, mutex, num_remaining, ready_promise, location_by_id, this, owner_address](
[2024-03-26T12:45:33Z] #13 160.1                                                                                       ^
[2024-03-26T12:45:33Z] #13 160.1     src/ray/core_worker/core_worker.cc:1826:34: error: reference to local binding 'owner_object_ids' declared in enclosing function 'ray::core::CoreWorker::GetLocationFromOwner'
[2024-03-26T12:45:33Z] #13 160.1               for (size_t i = 0; i < owner_object_ids.size(); ++i) {
[2024-03-26T12:45:33Z] #13 160.1                                      ^
[2024-03-26T12:45:33Z] #13 160.1     src/ray/core_worker/core_worker.cc:1810:36: note: 'owner_object_ids' declared here
[2024-03-26T12:45:33Z] #13 160.1       for (const auto& [owner_address, owner_object_ids] : objects_by_owner) {
[2024-03-26T12:45:33Z] #13 160.1                                        ^
[2024-03-26T12:45:33Z] #13 160.1     src/ray/core_worker/core_worker.cc:1829:19: error: reference to local binding 'owner_object_ids' declared in enclosing function 'ray::core::CoreWorker::GetLocationFromOwner'
[2024-03-26T12:45:33Z] #13 160.1                       owner_object_ids[i], std::make_shared<ObjectLocation>(locations[i]));
[2024-03-26T12:45:33Z] #13 160.1                       ^
[2024-03-26T12:45:33Z] #13 160.1     src/ray/core_worker/core_worker.cc:1810:36: note: 'owner_object_ids' declared here
[2024-03-26T12:45:33Z] #13 160.1       for (const auto& [owner_address, owner_object_ids] : objects_by_owner) {
[2024-03-26T12:45:33Z] #13 160.1                                        ^
[2024-03-26T12:45:33Z] #13 160.1     src/ray/core_worker/core_worker.cc:1833:31: error: reference to local binding 'owner_address' declared in enclosing function 'ray::core::CoreWorker::GetLocationFromOwner'
[2024-03-26T12:45:33Z] #13 160.1                                << owner_address.worker_id()
[2024-03-26T12:45:33Z] #13 160.1                                   ^
[2024-03-26T12:45:33Z] #13 160.1     src/ray/core_worker/core_worker.cc:1810:21: note: 'owner_address' declared here
[2024-03-26T12:45:33Z] #13 160.1       for (const auto& [owner_address, owner_object_ids] : objects_by_owner) {
[2024-03-26T12:45:33Z] #13 160.1                         ^
[2024-03-26T12:45:33Z] #13 160.1     src/ray/core_worker/core_worker.cc:1821:77: error: lambda capture 'this' is not used [-Werror,-Wunused-lambda-capture]
[2024-03-26T12:45:33Z] #13 160.1         [owner_object_ids, mutex, num_remaining, ready_promise, location_by_id, this, owner_address](
[2024-03-26T12:45:33Z] #13 160.1                       
```

File: src/ray/core_worker/core_worker.cc (hunk 5 above)
Comment (Bye-legumes) [reply src/ray/core_worker/core_worker.cc]: changed and remove CreateObjectLocations function

**Comment by Bye-legumes** [comment]
Currently the mutable_object_test  is flaky(I found others ci also timeout on this test or flaky) and I don't think my modification will have influences on the plasma level as the mutable_object_test  use the api from plasma level not core_worker level, right?

**Comment by Bye-legumes** [comment]
@raulchen Here is the results as I can only test with 2e5 as that will take 32GB memory.
Screenshot

File: src/ray/core_worker/core_worker.cc [hunk 7]
```diff
@@ -1782,38 +1782,76 @@ Status CoreWorker::GetLocationFromOwner(
...
-          } else {
-            RAY_LOG(WARNING) << "Failed to query location information for " << object_id
-                             << " with error: " << status.ToString();
-          }
-          (*num_remaining)--;
-          if (*num_remaining == 0) {
-            ready_promise->set_value();
-          }
-        });
+  for (const auto &pair : objects_by_owner) {
```
Comment (jjyao) [inline src/ray/core_worker/core_worker.cc:1798]: nit: pair -> owner_and_objects

File: src/ray/core_worker/core_worker.cc [hunk 8]
```diff
@@ -1782,38 +1782,76 @@ Status CoreWorker::GetLocationFromOwner(
...
+    const auto &owner_address = pair.first;
+    const auto &owner_object_ids = pair.second;
+
+    // Calculate the number of batches
+    // Use the same config from worker_fetch_request_size
+    int64_t batch_size = RayConfig::instance().worker_fetch_request_size();
+    size_t num_batches =
+        (owner_object_ids.size() + batch_size - 1) / batch_size;  // round up
+    *num_remaining += num_batches;
```
Comment (jjyao) [inline src/ray/core_worker/core_worker.cc:1807]: we can just increment num_remaining inside the for-loop?

File: src/ray/core_worker/core_worker.cc [hunk 9]
```diff
@@ -3772,15 +3812,17 @@ void CoreWorker::HandleGetObjectLocationsOwner(
...
-  auto object_info = reply->mutable_object_location_info();
-  auto status = reference_counter_->FillObjectInformation(object_id, object_info);
+  Status status = Status::OK();  // Assume success initially
+  for (int i = 0; i < request.object_ids_size(); ++i) {
+    auto object_id = ObjectID::FromBinary(request.object_ids(i));
+    auto object_info = reply->add_object_location_infos();
+    // TODO(zhilong): Need to deal with fail query for each object?
+    reference_counter_->FillObjectInformation(object_id, object_info);
+  }
   send_reply_callback(status, nullptr, nullptr);
```
Comment (jjyao) [inline src/ray/core_worker/core_worker.cc:3826]: nit: can just be `send_reply_callback(Status::OK(), nullptr, nullptr)`

File: src/ray/core_worker/core_worker.cc [hunk 10]
```diff
@@ -3772,15 +3812,17 @@ void CoreWorker::HandleGetObjectLocationsOwner(
...
     return;
   }
-  auto object_id = ObjectID::FromBinary(object_location_request.object_id());
-  auto object_info = reply->mutable_object_location_info();
-  auto status = reference_counter_->FillObjectInformation(object_id, object_info);
+  Status status = Status::OK();  // Assume success initially
+  for (int i = 0; i < request.object_ids_size(); ++i) {
+    auto object_id = ObjectID::FromBinary(request.object_ids(i));
+    auto object_info = reply->add_object_location_infos();
+    // TODO(zhilong): Need to deal with fail query for each object?
```
Comment (jjyao) [inline src/ray/core_worker/core_worker.cc:3823]: what do you mean?

**Review by jjyao — COMMENTED** [review]
LG!

**Comment by jjyao** [comment]
`mutable_object_test` is unrelated.

**Review by jjyao — APPROVED** [review]

File: src/ray/core_worker/core_worker.cc (hunk 7 above)
Comment (Bye-legumes) [reply src/ray/core_worker/core_worker.cc]: fixed!

File: src/ray/core_worker/core_worker.cc (hunk 8 above)
Comment (Bye-legumes) [reply src/ray/core_worker/core_worker.cc]: fixed!

File: src/ray/core_worker/core_worker.cc (hunk 9 above)
Comment (Bye-legumes) [reply src/ray/core_worker/core_worker.cc]: fixed!

File: src/ray/core_worker/core_worker.cc (hunk 10 above)
Comment (Bye-legumes) [reply src/ray/core_worker/core_worker.cc]: This is originally I add error handle here then we just removed that. But I forgot to also delete this remark. Deleted now!

**Comment by Bye-legumes** [comment]
@jjyao  resolved all comments and passed CI. Ready to merge.

**Comment by Bye-legumes** [comment]
It seems the CI can pass this time. That great! Thanks! @jjyao

**Pull request merged**